Set homepage in templates when needed

reneeotten opened the pull request:
If upstream has set a homepage, and that one is different than what is set by the ```perl5-1.0``` PG, add it to the template. Otherwise let the PG take care of it.

Steap commented:
I know I wrote some "complex things" in the templates (such as the "depends" macro), but I usually like to keep the templates simple for two  reasons:

- it makes them easier to read
- it is easier to test Python code than Jinja2 templates (by which I mean "I can't be bothered to find a proper way of testing templates" :p)

So I would rather do something like this:

```python
class MacportsPackage(object):
    ...
    @property
    def homepage(self):
        return self.upt_pkg.homepage

class MacPortsPerlPackage(MacPortsPackage):
    ...
    @property
    def homepage(self):
        if foo:
            return 'http://...'
        else:
            return None
```

In the template we would write:

```jinja
{% if pkg.homepage %}
homepage            {{ pkg.homepage }}
{% endif %}
```

We could then fix the Rubygems issue the same way, and have a bunch of unit tests for everything.

reneeotten commented:
@Steap I see your point and do think it makes sense to do some of this in the backend instead of templates.

In fact, I started to add a function to automtically "convert" upstream homepages to ```https``` if we know that the hosts support them (e.g., github.com, rubygems.org, etcetera). That is indeed in the ```MacportsPackage``` already, so let me try and combine the PRs with ```homepage```-related changes into one.

Steap commented:
Do you have examples of packages that report "http" instead of "https"? I think this change should make its way to the frontends, as all backends would probably like that.

reneeotten commented:
okay, I will update this PR to do the homepage conversion stuff in the backend and simplify the templates.

> Do you have examples of packages that report "http" instead of "https"? I think this change should make its way to the frontends, as all backends would probably like that.

I agree, I think it would actually make sense to add this to ```upt``` itself as not to duplicate the code in every frontend. Something along the lines I added to ```upt_macports``` would probably work:
```
    def _homepage(self):
        upstream_homepage = self.upt_pkg.homepage

        if upstream_homepage.startswith('https'):
            return upstream_homepage

        https_hosts = ('github', 'readthedocs', 'pypi', 'bitbucket',
                       'metacpan', 'rubygems')
        homepage_split = upstream_homepage.lstrip('http://').split('.')

        if set(homepage_split).intersection(https_hosts):
            return upstream_homepage.replace('http', 'https')
        else:
            return upstream_homepage
```

Below are a few examples of ports for which upstream is using ```http```, but that could use ```https```.
- Python/upt-pypi: bigfloat, bottle-sqlalchemy, cffi, https-parser, python-mingus, py, pycares, s3fs, tox, tweepy
- Perl/upt-cpan: Astro-FITS-Header, Catalyst-Plugin-Authorization-ACL, Code-TidyAll, Crypt-OpenSSL-RSA, Data-Visitor
- Ruby/upt-rubygems: activeldap, audiofile, bones, enum, narray

Steap commented:
I went for a slightly different implementation, what do you think about https://framagit.org/upt/upt/commit/5409ff3a78b78eb526a109b15df3b9ae8c6d5028 ?

reneeotten commented:
> I went for a slightly different implementation, what do you think about https://framagit.org/upt/upt/commit/5409ff3a78b78eb526a109b15df3b9ae8c6d5028 ?

Yes, your way of checking if one can use ```https``` is much more general and easier to maintain than my poor-man's-approach of adding "known, good hosts that support https". Looks good to me!

I am flying out for the holidays in a few hours, but will eventually update this PR with a more generalized solution for setting the homepage.

reneeotten commented:
@Steap how about this?

Steap commented on upt_macports/upt_macports.py:
```diff
@@ -207,6 +228,14 @@ def _normalized_macports_folder(name):
+            return homepage
+        else:
+            return f'https://rubygems.org/gems/{self.upt_pkg.name}'
+
```
So, in the case of Ruby, using this default URL is already done in upt-rubygems (see https://framagit.org/upt/upt-rubygems/blob/master/upt_rubygems/upt_rubygems.py#L170). Obviously the maintainer could set the homepage to "http://dead-link.com", but hopefully that does not happen too often :)

reneeotten replied:
right, but currently if the metadata contains something random as ```unknown``` or whatever, it will still return that as URL... that's not what we want. It should be set to a default if it's not a valid URL.

Steap commented on upt_macports/upt_macports.py:
```diff
@@ -189,6 +201,15 @@ def _cpandir(self):
+        if homepage.startswith('http') and portgroup_default not in homepage:
+            return homepage
+        else:
+            return None
```
Why not use metacpan.org/release/{pkg} as the default value?

reneeotten replied:
well, that's not the value that is used as default in the Perl portgroup; so it if we decide this should be the default it will have to be changed there as well.

Steap commented on upt_macports/upt_macports.py:
```diff
@@ -147,6 +151,14 @@ def _normalized_macports_folder(name):
+        if homepage.startswith('http'):
+            return homepage
+        else:
+            return f'https://pypi.org/project/{self.upt_pkg.name}'
```
Honestly, I should put this default value in upt-pypi. What do you think?

reneeotten replied:
sure, that wouldn't hurt! But one should make sure then that all situations where the metadata doesn't contain a valid URL are set to this value.

reneeotten commented:
@Steap any additional comments here? If not, I would like to merge this. Of course it would be helpful to add a default in ```upt-pypi``` and such; once that has happened we can always update the code in ```upt-macports``` accordingly. Also, I will look into the URL you suggest for CPAN packages; if we go for that it will need to be changed in the ```perl``` PortGroup as well. Again we can defer this to a later time I think.

reneeotten commented:
okay, I am going to merge this now - we can make adjustments if things change in ```upt``` or the frontends.
